dds/DCPS/DiscoveryBase.h

```diff
@@ -957,6 +997,164 @@ namespace OpenDDS {
+            - the topic name
+            - the type name
+            - the type identifiers
+            - etc.
+          Then you are done.
+          Otherwise, we have to think about cases where a local/discovered changes in a way that alters consistency.
+          If we do have to support this, then it will have to be in future work.
+         */
+
+        // 1. collect type info about the writer, which may be local or discovered
+        XTypes::TypeInformation* writer_type_info = 0;
+
+        const LocalPublicationIter lpi = local_publications_.find(writer);
+        DiscoveredPublicationIter dpi;
+        bool writer_local = false, already_matched = false;
+        if (lpi != local_publications_.end()) {
+          writer_local = true;
+          writer_type_info = &lpi->second.type_info_;
+        } else if ((dpi = discovered_publications_.find(writer))
+                   != discovered_publications_.end()) {
+          writer_type_info = &dpi->second.type_info_;
+        } else {
+          return; // Possible and ok, since lock is released
+        }
+
+        // 2. collect type info about the reader, which may be local or discovered
+        XTypes::TypeInformation* reader_type_info = 0;
+
+        const LocalSubscriptionIter lsi = local_subscriptions_.find(reader);
+        DiscoveredSubscriptionIter dsi;
+        bool reader_local = false;
+        if (lsi != local_subscriptions_.end()) {
+          reader_local = true;
+          reader_type_info = &lsi->second.type_info_;
+        } else if ((dsi = discovered_subscriptions_.find(reader))
+                   != discovered_subscriptions_.end()) {
+          reader_type_info = &dsi->second.type_info_;
+        } else {
+          return; // Possible and ok, since lock is released
+        }
+
+        MatchingData md;
+
+        // if the type object is not in cache, send RPC request
+        md.writer = writer;
+        md.reader = reader;
+        md.time_added_to_map = MonotonicTimePoint::now();
+
+        if ((writer_type_info->minimal.typeid_with_size.type_id.kind() != XTypes::TK_NONE) &&
+            (reader_type_info->minimal.typeid_with_size.type_id.kind() != XTypes::TK_NONE)) {
+          if (!writer_local && reader_local) {
+            if (type_lookup_service_ && !type_lookup_service_->type_object_in_cache(writer_type_info->minimal.typeid_with_size.type_id)) {
+              XTypes::TypeIdentifierSeq type_ids;
+              type_ids.append(writer_type_info->minimal.typeid_with_size.type_id);
+              md.rpc_sequence_number = ++type_lookup_service_sequence_number_;
+              MatchingDataIter md_it = matching_data_buffer_.find(MatchingPair(writer, reader));
+              if (md_it != matching_data_buffer_.end()) {
+                md_it->second = md;
+              } else {
+                matching_data_buffer_.insert(std::make_pair(MatchingPair(writer, reader), md));
+              }
```

Comment:
This was factored out into the `save_matching_data_and_get_typeobjects` function, so it should be removed.